### ISSUE TITLE
Flickering when toggling nvim-tree or neotest

### ISSUE THREAD

**Issue by sand4rt** [shortcuts/no-neck-pain.nvim#258]
## Description

https://github.com/shortcuts/no-neck-pain.nvim/assets/17591696/c11055c4-fce9-41e1-b917-1e831e49ec01

## Steps to reproduce

1. enable no-neck-pain with `:NoNeckPain`
2. open nvim-tree with `:NvimTreeToggle`
3. hide nvim-tree with `:NvimTreeToggle`

The window flickers 

## Expected behavior

The window should not flicker

## Environment

- Neovim version: 0.9.x
- no-neck-pain.nvim version: latest
- Plugin clash: nvim-tree, neotest

**Comment by shortcuts** [comment]
Hey thanks for reporting (and still using the plugin :D)

I assume the debounce is too high so the resize is not instantaneous, I'll try to tweak it!

If you want to give a try to tweaking it, it's [this variable](https://github.com/shortcuts/no-neck-pain.nvim/blob/2bcb6b761a34c69739da9aab642839b59236b801/lua/no-neck-pain/util/api.lua#L66) that controls the timeout

**Comment by sand4rt** [comment]
Hi @shortcuts thanks for your reply and plugin :)

When I set the `timeout` to 0-10 it seems like it is a little less, but still quite visible :(

**Comment by sand4rt** [comment]
Hi @shortcuts, do you have any pointers on how to solve this annoying issue? Tweaking the timeout did not work unfortunately. The plugin would be perfect once this is resolved..

**Comment by shortcuts** [comment]
> Hi @shortcuts, do you have any pointers on how to solve this annoying issue? Tweaking the timeout did not work unfortunately. The plugin would be perfect once this is resolved..

Hey, thanks for the feedback, glad you still like the plugin :D I've been in parental leave for a few months now so I may be slower to answer, sorry!

I need to come back to this indeed! While developing the integration feature I tried to make the buffer placement super consistent but it has the downside of this flickering, I'll try to see if I can improve it but I'm not sure how it will impact consistency

**Comment by shortcuts** [comment]
Let me know if 30 is still too annoying for you, I need to find a workaround because TSPlayground impacts other integration which forces us to have >5ms timeout :/

**Comment by sand4rt** [comment]
> Hey, thanks for the feedback, glad you still like the plugin :D I've been in parental leave for a few months now so I may be slower to answer, sorry!

Congratulations! No worries, glad you responded! 

Adjusting the timeout makes it slightly better but there is not much of a difference. Completely removing the debounce function seems necessary to eliminate the flickering.

Probably not the best solution, but perhaps it is an option to give the user control over when NNP opens and closes? So there could be a keymap like:

```lua
vim.keymap.set("n", "<leader>e", function()
    vim.cmd.NoNeckPainToggleLeftSide()
    vim.cmd.Neotree("toggle")
end, { desc = "Toggle Neotree" })
```

I don't have much experience with Lua, but if I could help out with testing, please let me know!

**Comment by shortcuts** [comment]
thanks @sand4rt, actually I think I've found the root cause of the race conditions with tree-sitter playground, https://github.com/shortcuts/no-neck-pain.nvim/pull/311 should close this issue but feel free to re-open if you observe the issue again after upgrading to 1.8.3

**Comment by sand4rt** [comment]
Hi @shortcuts, thanks a lot again for looking into this. I tested the latest changes, but I don't see a difference unfortunately.

```lua
return {
    "shortcuts/no-neck-pain.nvim",
    branch = "release-please--branches--main--components--no-neck-pain.nvim",
    lazy = false,
    keys = {
        { "<Leader>z", ":NoNeckPain<CR>", noremap = true, silent = true },
    },
    opts = {
        width = 140,
        autocmds = {
            enableOnVimEnter = true,
        },
        mappings = {
            enabled = false,
        },
        integrations = {
            NeoTree = {
                position = "left",
                reopen = true,
            },
            undotree = {
                position = "left",
            },
            neotest = {
                position = "right",
                reopen = true,
            },
            TSPlayground = {
                position = "right",
                reopen = true,
            },
            NvimDAPUI = {
                position = "right",
                reopen = true,
            },
        },
    },
}
```

**Comment by shortcuts** [comment]
> Hi @shortcuts, thanks a lot again for looking into this. I tested the latest changes, but I don't see a difference unfortunately.
> 
> ```lua
> return {
>     "shortcuts/no-neck-pain.nvim",
>     branch = "release-please--branches--main--components--no-neck-pain.nvim",
>     lazy = false,
>     keys = {
>         { "<Leader>z", ":NoNeckPain<CR>", noremap = true, silent = true },
>     },
>     opts = {
>         width = 140,
>         autocmds = {
>             enableOnVimEnter = true,
>         },
>         mappings = {
>             enabled = false,
>         },
>         integrations = {
>             NeoTree = {
>                 position = "left",
>                 reopen = true,
>             },
>             undotree = {
>                 position = "left",
>             },
>             neotest = {
>                 position = "right",
>                 reopen = true,
>             },
>             TSPlayground = {
>                 position = "right",
>                 reopen = true,
>             },
>             NvimDAPUI = {
>                 position = "right",
>                 reopen = true,
>             },
>         },
>     },
> }
> ```

oh 1.8.3 isn't released yet, I'm waiting for a last fix to be done!

**Comment by shortcuts** [comment]
ah just saw you were testing from the release branch D:

**Comment by sand4rt** [comment]
@shortcuts is there any chance this could be reopened? There seems to be a slight improvement, but not much of a difference unfortunately.

**Comment by shortcuts** [comment]
> @shortcuts is there any chance this could be reopened?

Yes definitely

> There seems to be a slight improvement, but not much of a difference unfortunately.

It's worth mentioning that we can't really re-arrange buffer, so far I did not found a smoother way than closing the integration, enabling nnp, then re-opening it, which causes a small flickering

I'll try to see if that's something I can improve!

**Comment by shortcuts** [comment]
@sand4rt 

https://github.com/shortcuts/no-neck-pain.nvim/assets/20689156/abefae30-94c4-4ae5-8c05-62a626b5aad3

is this what you observe or it appears slower/faster on your screen? I guess this varies based on the hardware so it's hard to determine what's an acceptable flickering on my side

**Comment by shortcuts** [comment]
Hey @sand4rt it would be nice if you could give a shot to https://github.com/shortcuts/no-neck-pain.nvim/pull/319, now the only flickering remaining would be the resize of the window which is unavoidable sadly

**Comment by sand4rt** [comment]
Hey @shortcuts, thanks a lot again for looking into this! Flickering when resizing is ok with me as I don't resize the window's as often as I toggle neotree and neotest. I don't see a difference with #319 unfortunately:

https://github.com/shortcuts/no-neck-pain.nvim/assets/17591696/95f1d90c-5262-4352-b88f-db3df45848ff

It seems that when neotree or neotest is closed, the main buffer attempts to claim that space and then NNP steps in to reclaim that space again after a short delay. I think the only way to solve this is to remove the short delay/debounce function, as NNP must kick in immediately after neotest/neotree is closed otherwise the flickering appears. I tested this by adjusting the NNP code a bit to remove the debounce function and the flickering went away. _(I got loads of errors and race conditions tho as im not that familiar with Lua/VIM)_.

I've noticed the flickering in your video too. If you have a keymap for toggling neotree faster, you will observe the flickering more clearly. In my situation, having auto-close enabled on neotree is really annoying as I open neotree and then select a file right after, which loses neotree again.

**Comment by shortcuts** [comment]
Hey @sand4rt, I've made an other update to https://github.com/shortcuts/no-neck-pain.nvim/pull/319 which removes even more debounce.

I did some tests and even without debounces you'd still observe a flickering, the reason is that the side buffers are resized when the layout shifts (e.g. a new buffer gets added, a side tree in that case). My guess is that after altering the neovim layout, a redraw of the ui is queued, if an other layout shift appends during this redraw, it gets queued again or something similar, because toggling once doesn't create flickering, but toggling a few times in a row does (I used a keymap this time).

I hope what I've proposed in 319 doesn't ruin your user experience

**Comment by sand4rt** [comment]
Hey @shortcuts, thanks a lot again for looking into this! I just tested it but it did not make a difference sadly. 

I noticed that the layout does not flicker when neotree opens and is smaller than the NNP buffer, but I would like to keep the width equal to NNP. 

> but toggling a few times in a row does (I used a keymap this time)

Less rapid toggling also causes the flickering:

https://github.com/shortcuts/no-neck-pain.nvim/assets/17591696/06775acd-5c24-4542-8c1f-b9a0a514a384

Well, if it's unfixable then i'll keep using NNP as is for a couple more weeks and see whether I can get used to it
